### ISSUE TITLE
Need job priority configuration of jobs in a flow

### ISSUE THREAD

**Issue by vikramkone** [azkaban/azkaban#551]
Is there a way we can specify priority for each job in a flow? so that P0 jobs executed first before P1 jobs? I know that we can model prioroties as dependency chain, but a configuration setting in the job file would be best. 
Our scenario is that we have multiple azkaban jobs scheduled to run for backfilling data, and executing all jobs at once in parallel will starve the system. We want the latest jobs to run first and then older jobs etc. Basically, a priority queue needs to be implemented in the scheduler

**Comment by davidzchen** [comment]
I don't think there is currently a way to do this but what are the main downsides of modeling this via dependencies compared to setting priorities?

**Comment by juhoautio** [comment]
We have implemented this to manage parallel workload against a database. I'm thinking making a PR, but it doesn't seem like new features are getting into master easily, so I wonder if it would be worth the effort.

In `azkaban.properties` we define the groups.
```properties
# individual priority groups
prioritygroup.groups.test_group_analytics.maxParallelExecutions=2
prioritygroup.groups.test_group_etl.maxParallelExecutions=4
```

So this defines two global priority groups: `test_group_analytics` and `test_group_etl`.

It would be better to have these configured from the UI and stored in the database instead of props, but this was the easiest way to get it done as we rarely need to add new groups. Obviously azkaban executor node needs to be restarted to apply changes in group config.

Then in `.job` files we set
```properties
type=what_ever_job_type
prioritygroup.name=test_group_analytics
prioritygroup.priority=5
```

Now if we have for example 3 jobs using `test_group_analytics` with priorities 1, 5, 10 that all start at the same time, jobs with 1 & 5 will be started and 10 will wait for a free slot. Also if a job with higher priority (let's say 2) is launched while 1 & 5 are running, it will take the first free slot and the job with priority 10 will wait further.

Jobs are blocked in a similar fashion as with `concurrentOption=pipeline`, showed as QUEUED in the UI.

The way we did this acts globally, so it doesn't matter if the jobs are from different flows, executions, or even projects. If you would need to have it on flow level only, then a straight-forward extension would be to allow creating ad hoc priority groups in the `.job` file of the flow (similar conf as now in `azkaban.properties`) that would only live during the flow execution.

**Comment by juhoautio** [comment]
> what are the main downsides of modeling this via dependencies compared to setting priorities?

TLDR; With dependencies the child jobs are not run if any parent fails. Also there's more waiting and idle time when jobs can't be parallelized.

--
#### Scenarios to consider

Consider a flow with some parallel jobs like this:
```
A1 -> B1
A2 -> B2
A3 -> B3
```

Let's say
- We want to run all A jobs in parallel
- We want a limited number of B jobs to run at the same time
- We want to run each `An -> Bn` even if some other n chain fails
- We want to B jobs in priority order if there's a need to decide which one to run
- We don't want the resource for B jobs to be idle if there's any finished A job

You suggested something like this I suppose:
```
A1 -> B1 -> A2 -> B2 -> A3 -> B3
```
Problems with this:
- A jobs not run in parallel – takes more time to finish
- Can't have max 2 B jobs running in parallel, just one
- 2 & 3 not run if 1 fails etc.
- Lot of idle time for the B job resource

Another approach would be to have it like this
```
(A1, A2, A3) -> B1 -> B2 ->B3
```
This has same problems as the previous graph, except that A jobs can be run in paralle, but then again none of the B jobs can start before all A jobs have completed.

**Comment by juhoautio** [comment]
@logiclord made [#447 Enhance flow runner to prioritize ready jobs using some property]( https://github.com/azkaban/azkaban/issues/447) (the commit: https://github.com/logiclord/azkaban/commit/52afa732741d404074fb5b960e58ba08cd3117e8).

I don't quite get what it's supposed to do. Just ordering the jobs before submission doesn't seem too useful to me, because they will still all start executing immediately.

**Comment by logiclord** [comment]
@juhoautio My change was also on the similar lines but it catered to control execution order of jobs. You approach is much more general.

@HappyRay @georgezhlw @evlstyle can you please help @juhoautio ?

**Comment by juhoautio** [comment]
@logiclord thanks for comment, what did you mean by helping me?

What benefit do you get from controlling the execution order, when it's actually just the start order, if the jobs are parallel in the flow graph? Obviously if the flow is sequential then the ordering doesn't matter any way.

**Comment by logiclord** [comment]
@juhoautio actually there is a limit to number of jobs that can run in parallel (https://github.com/azkaban/azkaban/blob/master/azkaban-execserver/src/main/java/azkaban/execapp/FlowRunnerManager.java#L105). This change was made due to an internal feature request where users want to control order or execution to reduce flow run time.

**Comment by hanip** [comment]
Wow. This is exactly what i need right now. It will solve the problem I posted here, http://stackoverflow.com/q/41115795/3988676 

And after take some look, I think the fork @logiclord made may do the trick. Can you please explain more about the changes, how to use "job.priority"? And is it also working on flow property?

To be specific, if I have this on my D.job :
```
type=command
dependencies=A.flow,B,C
```
How to use the "job.priority" to make the B & C jobs are started before A.flow?

If you may, answering the stackoverflow would be very nice.

**Comment by yanghaoAndroid** [comment]
@logiclord  @juhoautio    very exellent idea，but the code of the master can support this funnction ? i  am very desire this  " job priority "   function .

**Comment by juhoautio** [comment]
@djaiswal83 what do you think about this:
https://github.com/azkaban/azkaban/issues/551#issuecomment-224554085
?
